contracts/mixnet/src/transactions.rs

```diff
@@ -512,6 +646,22 @@ pub(crate) fn try_remove_delegation_from_gateway(
             }
             .into()];
 
+            // update total_delegation of this node
+            let mut gateways_bucket = gateways(deps.storage);
+            // in some rare cases the gateway bond might no longer exist as the node unbonded
+            // before delegation was removed. that is fine
+            if let Some(mut existing_bond) =
+                gateways_bucket.may_load(gateway_identity.as_bytes())?
+            {
+                // we should NEVER underflow here, if we do, it means we have some serious error in our logic
+                existing_bond.total_delegation.amount = existing_bond
+                    .total_delegation
+                    .amount
+                    .checked_sub(delegation)
+                    .unwrap();
+                gateways_bucket.save(gateway_identity.as_bytes(), &existing_bond)?;
+            }
+
```

Comment:
This (and the rest of the function) is pretty much duplicated code from the `try_remove_delegation_from_mixnode` function above. Maybe we should consider putting the mixnodes and gateways more and more together under some sort of inheritance pattern in the future. But probably now's not the time for it.

Reply:
Yeah, gateways and mixnodes have a lot of duplicate code, maybe we should do something about that in the future. Though, I'm slightly hesitant towards it as back in the past when we had centralised directory server, we've been doing exactly that and it's been a pain after a while